### PR TITLE
Implementation of /proposal route and proposal test cases

### PR DESCRIPTION
**Pull request by AuraOfDivinity**
## **Problem**
Github Issue Number: #130 

### **Solution of problem**
Completed implementation of the /proposal route and test cases. (2 Remaining test cases for file attachment and fetching org related proposals.)

**Generated docs** - [Link](https://documenter.getpostman.com/view/7312043/SzzdCLPr)

### **Type of Change**
[ ] Bug fix
[x] New Feature
[ ] Development of UI/UX prototypes
[ ] Small refactor
[ ] Change in Documentation

### **Checklist**
[x] My code follows the same style as the codebase
[x] My Code change requires a change in documentation
[x] I have updated the Readme accordingly
[x] I made PR against **development branch**
[x] I have run the test cases locally and it's passing.
[x] I have squashed my commits

### CODE REVIEW

File: app/controllers/proposal.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+          description: req.body.description,
+          fileLink: s3FileURL + file.originalname,
+          s3_key: params.Key
+        }
+        ProposalModel.updateOne(
+          { _id: proposalId },
+          { $push: { attachments: newFileUploaded } }
+        ).then((proposal) => {
+          console.log(proposal)
+        })
```
[the remark this answers is not in the document]
Comment (AuraOfDivinity) [reply app/controllers/proposal.js:74]: @devesh-verma I'm pushing the details of the uploaded file(i.e file name, link to access the file and the s3 key for the file) into the attachments attribute of the given proposal. Is there a better way I could have done this in your opinion Devesh?

File: app/models/Proposal.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      validate (title) {
+        if (validator.isEmpty(title)) {
+          throw new Error('Proposal Title cannot be kept empty')
+        }
+      }
+    },
+    organization: {
+      type: Schema.Types.ObjectId,
+      ref: 'Organization'
+    },
```
[the remark this answers is not in the document]
Comment (AuraOfDivinity) [reply app/models/Proposal.js:19]: Makes sense @devesh-verma  I'll remove it in the next commit.

File: app/controllers/proposal.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    } catch (error) {
+      HANDLER.handleError(res, error)
+    }
+  },
+
+  // Changes the state of a given proposal
+  changeState: async (req, res, next) => {
+    const proposalId = req.body.proposalId
+    const userId = req.body.userId
+    const proposalStatus = req.body.proposalStatus
```
[the remark this answers is not in the document]
Comment (devesh-verma) [reply app/controllers/proposal.js:114]: @AuraOfDivinity what about this?

**Comment by devesh-verma** [comment]
@AuraOfDivinity please remove the merge confict

File: .env.dev [hunk 4]
```diff
@@ -1,6 +1,10 @@
 PORT=5000
 NODE_ENV="development"
 JWT_SECRET="thisismysupersecrettokenjustkidding"
-DATABASE_URL="mongodb://localhost:27017/donut-development"
+DATABASE_URL="mongodb+srv://asel123:asel123@socialmediaappcluster-leoir.mongodb.net/example"
```
Comment (Rupeshiya) [inline .env.dev:4]: please change this to local, I think you forgot to remove your personal credentials

File: app.js [hunk 5]
```diff
@@ -1,74 +1,88 @@
...
+const organizationRouter = require("./app/routes/organisation");
+const commentRouter = require("./app/routes/comment");
+const projectRouter = require("./app/routes/project");
+const notificationRouter = require("./app/routes/notification");
+const proposalRouter = require("./app/routes/proposal");
 
-const app = express()
-const server = require('http').Server(app)
+const app = express();
+const server = require("http").Server(app);
```
Comment (Rupeshiya) [inline app.js:26]: Also, your code is violating, esLint rules, as we are using single quotes and no semicolon throughout the whole project. If CI would run then the test cases will fail.

File: app/controllers/proposal.js [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+      res.status(HttpStatus.CREATED).json({ proposal });
+    } catch (error) {
+      HANDLER.handleError(res, error);
+    }
+  },
+
+  // Updates the content of the proposal
+  saveProposal: async (req, res, next) => {
+    const { proposalId } = req.params;
+    const content = req.body.content;
```
Comment (Rupeshiya) [inline app/controllers/proposal.js:52]: just use destructuring for better flexibility

File: app/models/Proposal.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+const mongoose = require("mongoose");
+const Schema = mongoose.Schema;
+const validator = require("validator");
+
+const proposalSchema = new Schema(
+  {
+    title: {
```
Comment (Rupeshiya) [inline app/models/Proposal.js:7]: title and content is optional??

File: test/proposal.test.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+})
+
+// Testing proposal update
+test('Should update the content of the proposal', async (done) => {
+  await request(app)
+    .patch(`/proposal/${testProposalId}`)
+    .set('Authorization', `Bearer ${token}`)
+    .send(updatedProposalContent)
+    .expect(HttpStatus.OK)
+
```
Comment (Rupeshiya) [inline test/proposal.test.js:162]: check if the proposal collections is updated or not?

File: test/proposal.test.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+  proposalId: testProposalId
+}
+
+test('Should delete the proposal', async (done) => {
+  await request(app)
+    .delete('/proposal')
+    .set('Authorization', `Bearer ${token}`)
+    .send(deleteProposalContent)
+    .expect(HttpStatus.OK)
+
```
Comment (Rupeshiya) [inline test/proposal.test.js:177]: check if that proposal still exists in DB?

File: test/proposal.test.js (hunk 8 above)
Comment (AuraOfDivinity) [reply test/proposal.test.js:162]: @Rupeshiya Will be fixing this and the other issue on the bottom by Monday. I'm working on writing the remaining test cases as well.

**Comment by devesh-verma** [comment]
@AuraOfDivinity squash the commits so that I can merge this PR.

**Pull request merged**